docs/check.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	ctx := &checkContext{
+		categoryNames:  categoryNames,
+		docsPath:       docsPath,
+		knownFiles:     StringSet{},
+		markdownLinks:  map[string][]*relativeLink{},
+		markdownErrors: map[string][]string{},
+	}
+
+	// scan through markdown files
+	for _, langDir := range []string{"en", "cn"} {
+		if err := filepath.Walk(filepath.Join(docsPath, langDir),
+			func(p string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+				if info.IsDir() {
+					return nil
+				}
+				if strings.HasSuffix(info.Name(), ".md") {
+					if err := checkFile(p, ctx); err != nil {
+						return err
+					}
+				}
+				return nil
+			},
+		); err != nil {
+			return fmt.Errorf("error traversing through md files in %v: %v", filepath.Join(docsPath, langDir), err)
+		}
+	}
+	// scan through img and resources directories to update known files
+	for _, dir := range []string{"img", "resources"} {
+		if err := filepath.Walk(filepath.Join(docsPath, dir),
+			func(p string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+				if info.IsDir() {
+					return nil
+				}
+				ctx.knownFiles.Add(strings.TrimPrefix(p, docsPath))
+				return nil
+			},
+		); err != nil {
+			return fmt.Errorf("error traversing through files in %v: %v", filepath.Join(docsPath, dir), err)
+		}
+	}
+
+	ctx.checkLinks()
+
+	if len(ctx.markdownErrors) > 0 {
+		errLines := []string{"Errors found in documentation markdown"}
+		for f, errs := range ctx.markdownErrors {
+			errLines = append(errLines, fmt.Sprintf("  %v:", strings.TrimPrefix(f, repoRoot)))
+			for _, err := range errs {
+				errLines = append(errLines, fmt.Sprintf("    %s", err))
+			}
+		}
+		return fmt.Errorf("%v", strings.Join(errLines, "\n"))
+	}
+
+	return nil
+}
+
+// parseCategoryNames parses the given config file for the list of category names
+func parseCategoryNames(configPath string) (StringSet, error) {
+	f, err := os.Open(configPath)
+	if err != nil {
+		return nil, fmt.Errorf("error opening file at %v: %v", configPath, err)
+	}
+	defer f.Close()
+
+	const categoryListLine, categoryPrefix = "categoryList:", "  - "
+	categoryNames := StringSet{}
+
+	found := false
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		l := scanner.Text()
+		if l == categoryListLine {
+			found = true
+			continue
+		}
+		if found {
+			if !strings.HasPrefix(l, categoryPrefix) {
+				found = false
+				continue
+			}
+			categoryNames.Add(strings.TrimPrefix(l, categoryPrefix))
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error scanning file: %v", err)
+	}
+	return categoryNames, nil
+}
+
+var (
+	// general format of a relative link, where the link will be computed by a jekyll function encapsulated in {{ }}
+	relativeLinkRe = regexp.MustCompile(`\[.+\]\({{.*}}\)`)
+	// path encapsulated in ' ' could have an optional search query "?q=queryStr" and/or an optional anchor reference "#anchor"
```

Comment:
Is this only for links that link to other doc pages? 

For external links this doesn't capture all possible scenarios

Reply:
yes i do not want to validate any external links in a PR build